File: level1.c
#include <ncurses.h>
#include <unistd.h>

#include "level1.h"
#include "map.h"
#include "ui.h"

const char *level_1_map =
    "                  "
    "   #########      "
    "   #       ###### "
    "   #      0|    # "
    "   #       ###  # "
    "   # L     # #  # "
    "   ######### #--# "
    "             #### ";

static Map *map;

static void finish_level_1();

int event0(void *_map, int event_pos) {
    Map *map = (Map *)_map;
    clear();
    printw("Let's open the door!");
    refresh();
    sleep(1);
    printw("\nPress ANY key to continue.");
    getch();

    map->pos = event_pos;
    map->data[map->pos + 1] = ' ';
    map->data[map->pos] = ' ';
    return 0;
}

void run_level_1() {
    int ch, keep_playing = 1;
      
    map = map_create(level_1_map, "level 1 map", 18, 8);

    // set events
    map->events[0] = &event0;

    render_map(map);

    while(keep_playing == 1) {
        ch = getch();

        switch (ch) {
        case 'q':
        case 'Q':
        case 127: // DEL
            keep_playing = 0;
            break;
        default:
            map_move(map, ch);
            render_map(map);
            break;
        }
    }

    finish_level_1();
}

static void finish_level_1() {
    map_destroy(map);
}

File: src/level_1.rs
use ncurses::*;
use std::char;
use std::thread::sleep;
use std::time::Duration;
use map::*;

static LEVEL_1: &'static str =
    concat!("                  ",
            "   #########      ",
            "   #       ###### ",
            "   #      0|    # ",
            "   #       ###  # ",
            "   # L     # #  # ",
            "   ######### #--# ",
            "             #### ");

fn event0(m: &mut Map, event_pos: usize) -> bool {
    clear();
    printw("Let's open the door!");
    refresh();
    sleep(Duration::new(1,0));
    printw("\nPress ANY key to continue.");
    getch();

    m.pos = event_pos;
    // m.data[m.pos + 1] = ' ';
    // m.data[m.pos] = ' ';
    return true;
}

pub fn run_level_1() {
    // let mut map = Map::new(LEVEL_1.to_string(), 18, 8, vec![event0]);
    let mut map = map_from_str(LEVEL_1_);

    map.render();

    let mut keep_playing: bool = true;

    while keep_playing == true {
        let ich = getch() as u32;
        let maybe_ch = char::from_u32(ich);

        let ch = match maybe_ch {
            Some(ch) => ch,
            None     => {
                warn!("Unknown character: {}", ich);
                'q'
            }
        };

        if ch == 'q' || ch == 'Q' {
            keep_playing = false;
        } else {
            map.move_pos(ch);
            map.render();
        }
    }
}

const LEVEL_1_: &'static str = r"

 #########
 #       ########################
 #   L   |        |             #
 #       #######--####          #
 #########   #    #  #          #
             #    #  #####---####
             ######      #   #
                         #   #
             #######     #   #
             #     #######   #
             #     |         #
             #################
";
